### DISCUSSION TITLE
Support for Custom SMS Providers

### DISCUSSION THREAD

**Discussion by execreate** [zitadel/zitadel#5573] [Ideas]
The current implementation of Zitadel only supports Twilio as an SMS provider. This limits users who want to utilize other SMS providers or have a custom implementation for sending SMS messages.

I propose adding support for custom SMS providers in Zitadel by introducing a more flexible and extensible SMS provider integration. This can be achieved by:
1. Creating an SMS provider interface that defines the required methods to send SMS messages, such as send_sms(to, message).
2. Implementing the existing Twilio integration as a separate class that implements the new SMS provider interface.
3. Adding support for users to configure and use custom SMS providers that implement the same interface, through configuration files or environment variables.
4. Providing documentation and examples for implementing custom SMS providers, which will enable users to easily integrate their preferred SMS provider.

An alternative approach could be to add support for more popular SMS providers directly within Zitadel. However, this might not cover all use cases and may require frequent updates to accommodate new providers or changes in their APIs. By allowing users to implement custom SMS providers, we can ensure that Zitadel remains flexible and adaptable to any SMS provider without needing constant updates.

The ability to use custom SMS providers would greatly enhance the flexibility of Zitadel and accommodate the diverse needs of its users. By allowing users to integrate their preferred SMS provider, they can have more control over their costs, regional coverage, and other preferences specific to their use case. This feature would make Zitadel a more versatile and inclusive platform for organizations with different communication requirements.

**Comment by hifabienne** [comment]
Hei @execreate 

ZITADEL is built that we could extend it with different providers.
This means implementing each provider we want to enable.
At the moment the sms feature is not used a lot, as phone numbers are not required in ZITADEL.
So at the moment it is not on the roadmap to extend ZITADEL with more providers. But if we see that there is a high demand, we will have a look at it.
What kind of provider would you need?

**Reply by execreate** [reply]
Hey @hifabienne ,
Thanks for your quick reply! I was thinking about some unified set of parameters that can be used with any custom SMS provider. For instance:

- send message endpoint (https://my-sms-provider.com/send)
- sms provider authentication method and credentials (API key, username-password, etc.)
- request body template (which gets populated with a destination phone number, magic link, OTP code, etc.)

This setup can be useful for using local sms gateway providers like we do currently in our project.

**Comment by sevensolutions** [comment]
I'am also interested in such extensibility feature.
I think the easiest option would be to simply allow the registration of a webhook.
